django/db/backends/base/operations.py

```diff
@@ -714,42 +714,68 @@ def subtract_temporals(self, internal_type, lhs, rhs):
             "This backend does not support %s subtraction." % internal_type
         )
 
-    def window_frame_start(self, start):
-        if isinstance(start, int):
-            if start < 0:
-                return "%d %s" % (abs(start), self.PRECEDING)
-            elif start == 0:
+    def window_frame_value(self, value, none_value):
+        if isinstance(value, int):
+            if value < 0:
+                return "%d %s" % (abs(value), self.PRECEDING)
+            elif value == 0:
                 return self.CURRENT_ROW
-        elif start is None:
-            return self.UNBOUNDED_PRECEDING
-        raise ValueError(
-            "start argument must be a negative integer, zero, or None, but got '%s'."
-            % start
-        )
+            else:
+                return "%d %s" % (value, self.FOLLOWING)
+        elif value is None:
+            return none_value
+
+    def window_frame_rows_start(self, start):
+        as_sql = self.window_frame_value(start, none_value=self.UNBOUNDED_PRECEDING)
+        if as_sql is None:
+            raise ValueError(
+                "start argument must be an integer, zero, or None, but got '%s'."
+                % start
+            )
+        return as_sql
 
-    def window_frame_end(self, end):
-        if isinstance(end, int):
-            if end == 0:
-                return self.CURRENT_ROW
-            elif end > 0:
-                return "%d %s" % (end, self.FOLLOWING)
-        elif end is None:
-            return self.UNBOUNDED_FOLLOWING
-        raise ValueError(
-            "end argument must be a positive integer, zero, or None, but got '%s'."
-            % end
-        )
+    def window_frame_rows_end(self, end):
+        as_sql = self.window_frame_value(end, none_value=self.UNBOUNDED_FOLLOWING)
+        if as_sql is None:
+            raise ValueError(
+                "end argument must be an integer, zero, or None, but got '%s'." % end
+            )
+        return as_sql
+
+    def window_frame_range_start(self, start):
+        as_sql = self.window_frame_value(start, none_value=self.UNBOUNDED_PRECEDING)
+        if isinstance(start, int) and start > 0 or as_sql is None:
+            raise ValueError(
+                "start argument must be a negative integer, zero, or None, "
+                "but got '%s'." % start
+            )
+        return as_sql
```

Comment:
I'm not sure if these hooks are needed, I'd move the logic to the `window_frame_range_start_end()` (the same with `window_frame_rows_end()` and `window_frame_rows_start()`). As far as I'm aware, we don't need such a small granularity.

Reply:
Thank you, it looks much nicer now